File: spec-traits/src/main/java/bsp/traits/DocsPriorityTrait.java
package bsp.traits;

import software.amazon.smithy.model.node.Node;
import software.amazon.smithy.model.shapes.ShapeId;
import software.amazon.smithy.model.traits.AbstractTrait;
import software.amazon.smithy.model.traits.AbstractTraitBuilder;
import software.amazon.smithy.model.traits.TraitService;
import software.amazon.smithy.utils.SmithyBuilder;
import software.amazon.smithy.utils.ToSmithyBuilder;

public class DocsPriorityTrait extends AbstractTrait implements ToSmithyBuilder<DocsPriorityTrait> {

    public static final ShapeId ID = ShapeId.from("docs#docsPriority");

    private final int priority;

    private DocsPriorityTrait(Builder builder) {
        super(ID, builder.getSourceLocation());
        this.priority = builder.priority;
    }

    public int getPriority() {
        return this.priority;
    }

    @Override
    protected Node createNode() {
        return Node.from(priority);
    }

    public static Builder builder() {
        return new Builder();
    }

    @Override
    public SmithyBuilder<DocsPriorityTrait> toBuilder() {
        return builder().priority(priority).sourceLocation(getSourceLocation());
    }

    public static final class Builder extends AbstractTraitBuilder<DocsPriorityTrait, DocsPriorityTrait.Builder> {

        private int priority;

        public Builder priority(int priority) {
            this.priority = priority;
            return this;
        }

        @Override
        public DocsPriorityTrait build() {
            return new DocsPriorityTrait(this);
        }
    }

    public static final class Provider implements TraitService {

        @Override
        public ShapeId getShapeId() {
            return ID;
        }

        @Override
        public DocsPriorityTrait createTrait(ShapeId target, Node value) {
            int priority = value.expectNumberNode().getValue().intValue();
            return builder().priority(priority).sourceLocation(value).build();
        }
    }
}

File: spec-traits/src/main/kotlin/traits/DocsPriorityTrait.kt
package traits

import software.amazon.smithy.model.node.Node
import software.amazon.smithy.model.shapes.ShapeId
import software.amazon.smithy.model.traits.AbstractTrait
import software.amazon.smithy.model.traits.AbstractTraitBuilder
import software.amazon.smithy.model.traits.TraitService
import software.amazon.smithy.utils.SmithyBuilder
import software.amazon.smithy.utils.ToSmithyBuilder

class DocsPriorityTrait private constructor(builder: Builder) : AbstractTrait(ID, builder.sourceLocation),
    ToSmithyBuilder<DocsPriorityTrait> {
    val priority: Int

    init {
        priority = builder.priority
    }

    override fun createNode(): Node {
        return Node.from(priority)
    }

    override fun toBuilder(): SmithyBuilder<DocsPriorityTrait> {
        return builder().priority(priority).sourceLocation(sourceLocation)
    }

    class Builder : AbstractTraitBuilder<DocsPriorityTrait, Builder>() {
        var priority = 0
        fun priority(priority: Int): Builder {
            this.priority = priority
            return this
        }

        override fun build(): DocsPriorityTrait {
            return DocsPriorityTrait(this)
        }
    }

    class Provider : TraitService {
        override fun getShapeId(): ShapeId {
            return ID
        }

        override fun createTrait(target: ShapeId, value: Node): DocsPriorityTrait {
            val priority = value.expectNumberNode().value.toInt()
            return builder().priority(priority).sourceLocation(value)!!.build()
        }
    }

    companion object {
        val ID: ShapeId = ShapeId.from("traits#docsPriority")
        fun builder(): Builder {
            return Builder()
        }
    }
}
